File: ruby/stream.rb
require_relative "addressable"
require_relative "pointer"

class Stream
	attr_accessor :pos
	attr_reader :mem

	def initialize(memory, pos)
		@mem = memory
		@pos = pos
	end
	
	def read()
		puts "pos = 0x%x" % @pos
		ret = @mem.read(@pos, 1)[0]
		@pos += 1
		return ret
	end
	
	def back()
		@pos -= 1
	end
	
	def read_pointer(size)
		res = Pointer.new(@mem, @pos, size)
		@pos += size
		return res
	end
end

File: python/stream.py
from pointer import Pointer, PointerSigned


class Stream:
    def __init__(self, memory, pos):
        self.mem = memory
        self.pos = pos

    def read(self):
        ret = self.mem.read(self.pos, 1)[0]
        self.pos += 1
        return ret

    def back(self):
        self.pos -= 1

    def read_pointer(self, size):
        res = Pointer(self.mem, self.pos, size)
        self.pos += size
        return res

    def read_signed_pointer(self, size):
        res = PointerSigned(self.mem, self.pos, size)
        self.pos += size
        return res
